### PR TITLE
feat(quantic): QuanticResultsPerPage fixes and refactor tests

### PR DESCRIPTION
**Pull request by erocheleau**
SFINT-5851

## IN THIS PR:
- We are migrating E2E tests for the `quanticResultsPerPage` component from Cypress to Playwright and adding unit tests.
- We are also updating the behavior whenever an erroneous `initialChoice` or no `initialChoice` is passed to the component. We said 
> "The initial selection for the number of result per page. This should be part of the `choicesDisplayed` option. By default, this is set to the first value in `choicesDisplayed`."

But in reality, we were always defaulting to the value `10` whenever no value was passed,
And if the value was not within the `choicesDisplayed` we would completely crash Quantic initialization by throwing an error.

Instead now, if the value is invalid (such as not a number, missing or not part of the `choicesDisplayed`) we actually default to the first value in `choicesDisplayed`. Log an error in the console, but do not throw.

If the values in `choicesDisplayed` is invalid, we log an error in the console, and display the componentError, but no longer throw an error.

## UNIT TESTS:
image

## E2E TESTS:
image

### CODE REVIEW

**Comment by mmitiche** [comment]
@erocheleau to avoid issues with the versioning we should update the PR title to `test(quantic): ....` or `fix(quantic): ....` as this PR also fixes the behaviour of the component.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js [hunk 1]
```diff
@@ -36,9 +36,15 @@ export default class QuanticResultsPerPage extends LightningElement {
...
    * @type {number}
-   * @defaultValue `10`
+   * @defaultValue `the first value in 'choicesDisplayed'`
    */
-  @api initialChoice = 10;
+  @api
+  get initialChoice() {
+    return this._initialChoice;
+  }
+  set initialChoice(value) {
```
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:45]: in second thought I think re-validating this property each time it changes is not necessary because this property is only used once, it's used on the initial render of the component so validating in the beginning is enough.

However the property `choicesDisplayed ` needs to be re-validated each time its value change. cause this value is used by the component to display the choices, and the user could at any time after the connected callback provide a new value for this property. so having a setter for this property that validates its value could be a good idea. what do you think?

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    });
+  });
+
+  describe('controller initialization', () => {
+    beforeEach(() => {
+      mockSuccessfulHeadlessInitialization();
+      prepareHeadlessState();
+    });
+
+    it('should build the results per page and subscribe to state', async () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:149]:
```suggestion
    it('should build the results per page and search status controllers and subscribe to state', async () => {
```

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  });
+
+  describe('with a custom #choicesDisplayed option', () => {
+    beforeEach(() => {
+      console.error = jest.fn();
+    });
+
+    afterEach(() => {
+      cleanup();
+    });
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:200]: this should not be needed as this is already called in the higher describe.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        choicesDisplayed: customChoices.join(','),
+      });
+      await flushPromises();
+
+      expect(functionsMocks.buildResultsPerPage).toHaveBeenCalledTimes(1);
+      expect(functionsMocks.buildResultsPerPage).toHaveBeenCalledWith(
+        exampleEngine,
+        {
+          initialState: {
+            numberOfResults: 100,
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:213]: slightly better imo
```suggestion
            numberOfResults: customChoices[0],
```

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+      );
+      expect(resultsPerPage).not.toBeNull();
+
+      const resultsPerPageOptions = Array.from(
+        element.shadowRoot.querySelectorAll(selectors.resultsPerPageOption)
+      );
+      expect(resultsPerPageOptions).toHaveLength(customChoices.length);
+
+      resultsPerPageOptions.forEach((option, index) => {
+        expect(option.number).toEqual(customChoices[index]);
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:229]: also checking for the selected value:
```suggestion
        expect(option.number).toEqual(customChoices[index]);
        expect(option.selected).toBe(
          customChoices[index] === customChoices[0]
        );
```

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+      console.error = jest.fn();
+      resultsPerPageState = {
+        ...resultsPerPageState,
+        numberOfResults: customInitialChoice,
+      };
+    });
+
+    afterEach(() => {
+      cleanup();
+    });
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:269]: should not be needed as it is executed in the higher afterEach

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+
+      const initializationError = element.shadowRoot.querySelector(
+        selectors.componentError
+      );
+      expect(initializationError).toBeNull();
+
+      expect(console.error).toHaveBeenCalledTimes(1);
+      expect(console.error).toHaveBeenCalledWith(
+        `The "initialChoice" option value "${customInvalidInitialChoice}" is not included in the "choicesDisplayed" option "${defaultChoices}". Defaulting to the first value of the choices.`
+      );
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:321]: I think it would also be useful here to check that the first option is selected in the HTML,
something like:
```
      resultsPerPageOptions.forEach((option, index) => {
        expect(option.selected).toBe(index === 0);
      });
```

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:1]: a unit test is missing here that tests the behaviour of the component when an option is clicked, when an option is clicked the component receives an `quantic__select`  event and executes the `set` method of the results per page controller with the right parameter.

We should simulate sending this event, and make sure  the `set` method of the results per page controller is called with the right parameter.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/fixture.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  options: Partial<ResultsPerPageOptions>;
+};
+
+type QuanticResultsPerPageE2ESearchFixtures =
+  QuanticResultsPerPageE2EFixtures & {
+    urlHash: string;
+  };
+
+type QuanticResultsPerPageE2eInsightFixtures =
+  QuanticResultsPerPageE2ESearchFixtures & {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/fixture.ts:30]: we should use the core type here and not the search one as the insight type should not contain `urlHash `
```suggestion
  QuanticResultsPerPageE2EFixtures & {
```

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/pageObject.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+  get allResultsPerPageOptionsButtons(): Locator {
+    return this.component.locator(
+      resultsPerPageElementsSelectors.resultsPerPageOptionButton
+    );
+  }
+
+  get firstResultsPerPageOptionButton(): Locator {
+    return this.allResultsPerPageOptionsButtons.first();
+  }
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/pageObject.ts:31]: I think it's better if we keep our selector generic by accepting an index as parameter at it selects the option at the given index.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    test.describe('with default options', () => {
+      test('should load as expected', async ({resultsPerPage}) => {
+        await test.expect(resultsPerPage.component).toBeVisible();
+        test
+          .expect(await resultsPerPage.allResultsPerPageOptionsButtons.count())
+          .toBe(defaultChoices.length);
+        test
+          .expect(resultsPerPage.selectedResultsPerPageOption)
+          .toHaveText(defaultInitialChoice.toString());
+      });
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts:28]: This looks like a unit that has already been tested in the unit tests file, I would remove it.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+          choicesDisplayed: customChoices.join(','),
+        },
+      });
+
+      test('should load as expected', async ({resultsPerPage}) => {
+        await test.expect(resultsPerPage.component).toBeVisible();
+        test
+          .expect(await resultsPerPage.allResultsPerPageOptionsButtons.count())
+          .toBe(customChoices.length);
+      });
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts:71]: Same remark here, for me this is a unit test that you already did.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts (hunk 10 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts:28]: Yea I know it's there "twice", I thought it still made sense to make sure it loaded correctly, that way it tests LWS/Locker, etc.

If a more complex test were to fail, it would be harder to know what the issue was, but if we simply have a "does the component load" test, I think it's easier to know that something went really wrong :) 

We don't have to do that for all components, but I think it's useful to have both a UT and a e2e test if the component simply loads.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts (hunk 11 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/quanticResultsPerPage.e2e.ts:71]: See reply https://github.com/coveo/ui-kit/pull/4889#discussion_r1932392914

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:1]: Oh good catch, this wasn't covered in the e2e tests and there was not UT, good point I'll add it.

**Comment by erocheleau** [comment]
> @erocheleau to avoid issues with the versioning we should update the PR title to `test(quantic): ....` or `fix(quantic): ....` as this PR also fixes the behaviour of the component.

Very true, I forgot! thanks!

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:45]: Good point, since the choices are actually not headless state, it's possible someone could indeed chance those values without us having to re-create the component.

I actually don't know what happens to a component if you change its property after it has already been created and connected. If it re-connects it or simply re-render it.

I think it's unlikely that it happens in a real use-case, but I'll change it to cover any potential corner case.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js (hunk 5 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:229]: I didn't check that because in order to influence which value is selected, it has to go through headless.

Essentially, we have to mock the Headless state like this:

```
storeState = {
        resultsPerPage: {
          numberOfResults: 100,
        }
      };
```

In order to test that the `100` is actually the correct value selected.
It goes outside of the scope of the test that checks the option `choicesDisplayed` imo, as we already test that we're initializing the controller with the correct value in `numberOfResults` with this test:

```
expect(functionsMocks.buildResultsPerPage).toHaveBeenCalledWith(
        exampleEngine,
        {
          initialState: {
            numberOfResults: 100,
          },
        }
      );
```

But I can add a test to make sure the Headless initial state is correctly rendered however.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/fixture.ts (hunk 8 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/fixture.ts:30]: It's like that everywhere though xD

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/pageObject.ts (hunk 9 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/e2e/pageObject.ts:31]: Turns out I actually don't use it...

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js [hunk 12]
```diff
@@ -59,12 +65,18 @@ export default class QuanticResultsPerPage extends LightningElement {
...
   labels = {
     showNResultsPerPage,
   };
 
   connectedCallback() {
+    this.choices = this.parseChoicesDisplayed();
+    if (!this._initialChoice) {
+      this._initialChoice = this.choices[0];
+    }
+    this.validateInitialChoice();
```
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:79]: Seems like the best moment to do it, if we wait for the `initialize`, we have to wait until Quantic is in the process of being initialized in order to validate the params that are essentially already available.

I'd rather do it at connectedCallback, than wait later.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:45]: I updated it, re-review if you think it covers all possibilities.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js (hunk 7 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:365]: If the first option is selected depends entirely on our mock of Headless ResultsPerPage state, and not on internal state of the component.

I created a separate test to make sure the headless state was properly represented as selected.

**Review by SimonMilord — APPROVED** [review]
After addressing mehdi's concerns and comments, it seems good to me!

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:53]: Looks good to me 👍

**Review by mmitiche — APPROVED** [review]
Looking good, GG!
